### PR TITLE
Implement `gpu_kernel_multiple_outputs`

### PR DESCRIPTION
**Pull request by crcrpar**
This PR introduces a variant of `gpu_kernel` for functions that return multiple values with `thrust::tuple`.
With this I simplified `prelu_cuda_backward_share_weights_kernel`.

### Why using `thrust::tuple`?
Because `std::tuple` does not support `operator=` on device code which makes the implementation complicated.

### CODE REVIEW

**Review by zasdfgbnm — APPROVED** [review]
@ngimel This refactors the logics of TensorIterator offset calculation, 32bit indexing splitting, etc. of  prelu kernel into `gpu_kernel_multiple_outputs`, so that the codes in Activation.cu become cleaner without these logics. It is a gpu kernel that supports multiple outputs, but does not support dynamic casting for simplicity. It is supposed to cover some cases where `gpu_kernel` is not usable.

**Comment by z-a-f** [comment]
@zasdfgbnm, @ngimel I will import it to phabricator (want to try it out internally). Please, commandeer the diff is ready to land.

@crcrpar Please, resolve the merge conflicts whenever you have a chance

File: aten/src/ATen/native/cuda/MemoryAccess.cuh [hunk 1]
```diff
@@ -67,12 +67,12 @@ struct vectorized_load_helper {
...
 struct unroll_load_helper {
-  template <typename args_t, typename policy_t, typename offset_t>
-  static __device__ void apply(policy_t &self, args_t *args, offset_t offset, int j, int num_outputs) {
+  template <typename args_t, typename policy_t, typename offset_t, typename loader_t>
+  static __device__ void apply(policy_t &self, args_t *args, offset_t offset, loader_t loader, int j, int num_outputs) {
     using arg_t = std::tuple_element_t<arg_index, args_t>;
     // `data` hold the data_ptr for tensors [output, input0, input1, ...], so we
     // need a +1 offset to get the input
-    auto ptr = reinterpret_cast<arg_t *>(self.data[arg_index + num_outputs]) + offset[arg_index];
+    auto ptr = reinterpret_cast<arg_t *>(loader.template load<arg_t>(self.data[arg_index + 1], offset[arg_index], arg_index));
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/MemoryAccess.cuh]: wait, should it be `arg_index+num_outputs` instead of `+1` here?

Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/MemoryAccess.cuh]: Yes, you are right. And fixed.

**Comment by crcrpar** [comment]
@z-a-f Thanks. Looks like @zasdfgbnm fixed conflicts on behalf of me including applying the fix suggested by @ngimel.

Thank you @zasdfgbnm, @ngimel

**Comment by z-a-f** [comment]
@ngimel, Should I land this?

**Comment by zasdfgbnm** [comment]
@z-a-f This is not ready. Some cut-pastes are needed to make it build on ROCm.

**Pull request closed without merging**